### PR TITLE
Restrict Login Codes To Their Origin Browser

### PR DESCRIPTION
**Pull request by sampoder**
This will fix #3438. How it does it is that it generates a unique token when some signs in (this is the "browser token"), then it stores that token locally as a cookie in the browser as well as in the DB, then, when someone inputs a code it checks that the two match.

### CODE REVIEW

File: app/models/login_code.rb [hunk 1]
```diff
@@ -23,6 +25,8 @@
 #  fk_rails_...  (user_id => users.id)
 #
 class LoginCode < ApplicationRecord
+  has_encrypted :browser_token
+  blind_index :browser_token
```
Comment (northeastprince) [inline app/models/login_code.rb]: Why does this need to be encrypted if it's already random?

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -64,7 +64,10 @@ def login_code
 
     initialize_sms_params
 
-    resp = LoginCodeService::Request.new(email: @email, sms: @use_sms_auth, ip_address: request.ip, user_agent: request.user_agent).run
+    @browser_token = SecureRandom.urlsafe_base64
+    cookies.encrypted[:browser_token] = { value: @browser_token, expires: Time.current + 15.minutes }
```
Comment (northeastprince) [inline app/controllers/users_controller.rb]: Can this just be `signed`?

**Comment by northeastprince** [comment]
Before we implement this, I'm wondering if we should consider implementing and switching to login links instead. I think it's more secure anyway and easier to use.

File: app/models/login_code.rb (hunk 1 above)
Comment (sampoder) [reply app/models/login_code.rb]: Good point, I was just copying the session_token logic

File: app/controllers/users_controller.rb (hunk 2 above)
Comment (sampoder) [reply app/controllers/users_controller.rb]: Wdym?

Comment (northeastprince) [reply app/controllers/users_controller.rb]: A signed cookie instead of an encrypted one.

Comment (sampoder) [reply app/controllers/users_controller.rb]: we can do that!

Comment (sampoder) [reply app/controllers/users_controller.rb]:
```suggestion
    cookies.signed[:browser_token] = { value: @browser_token, expires: Time.current + 15.minutes }
```

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -154,7 +157,8 @@ def exchange_login_code
     user = UserService::ExchangeLoginCodeForUser.new(
       user_id: params[:user_id],
       login_code: params[:login_code],
-      sms: params[:sms]
+      sms: params[:sms],
+      browser_token: cookies.encrypted[:browser_token]
```
Comment (sampoder) [inline app/controllers/users_controller.rb]:
```suggestion
      browser_token: cookies.signed[:browser_token]
```

**Comment by YodaLightsabr** [comment]
@sampoder: what's the status of this PR? what do we have to do to move forward with it? It'd be great to get this out as one of our small wins in the next 5 weeks.

**Comment by sampoder** [comment]
Other than the conflicts, this PR is good to review from my end

**Comment by sampoder** [comment]
I now seem to be getting that error for all of those case, @YodaLightsabr if you're still not getting that error, what does your setup look like?

File: app/models/login_code.rb [hunk 4]
```diff
@@ -23,6 +25,8 @@
 #  fk_rails_...  (user_id => users.id)
 #
 class LoginCode < ApplicationRecord
+  has_encrypted :browser_token
```
Comment (garyhtou) [inline app/models/login_code.rb]: maybe we can use https://api.rubyonrails.org/classes/ActiveRecord/SecureToken/ClassMethods.html

or, moving the token generation logic into the model using a `before_validation`

**Review by garyhtou — CHANGES_REQUESTED** [review]

**Comment by sampoder** [comment]
check out #4405 which does this using the fancy rails method!

**Pull request closed without merging**